### PR TITLE
SEL32: Add IPU device support using pthreads on Linux and Windows.

### PR DESCRIPTION
**Pull request by AZBevier**
SEL32: Update makefile and SEL32.vsproj file to add IPU device.
SEL32: Update README.md file.
SEL32: Do a general code cleanup.

### CODE REVIEW

File: SEL32/sel32_chan.c [hunk 1]
```diff
@@ -80,12 +80,13 @@
 
 uint32  channels        = MAX_CHAN;             /* maximum number of channels */
 int     subchannels     = SUB_CHANS;            /* maximum number of subchannel devices */
-int     irq_pend        = 0;                    /* pending interrupt flag */
+//int     irq_pend        = 0;                    /* pending interrupt flag */
```
Comment (pkoning2) [inline SEL32/sel32_chan.c:83]: If you want code to go away, please delete it.  Commented out code (or ifdef-ed out) is for debugging, it shouldn't be in completed code.

File: SEL32/sel32_con.c [hunk 2]
```diff
@@ -259,43 +343,16 @@ t_stat con_srvo(UNIT *uptr) {
...
         sim_debug(DEBUG_CMD, dptr,
-            "con_srvo INCH unit %02x: CMD %08x cmd %02x incnt %02x u4 %02x\n",
-            unit, uptr->CMD, cmd, con_data[unit].incnt, uptr->u4);
+            "con_srvo INCH unit %02x: CMD %08x cmd %02x incnt %02x u4 %02x inch %06x in2 %06x\n",
+            unit, uptr->CMD, cmd, con_data[unit].incnt, uptr->u4, mema, M[mema>>2]);
 
         /* now call set_inch() function to write and test inch buffer addresses */
         /* 1-256 wd buffer is provided for 128 status dbl words */
-        tstart = set_inch(uptr, mema, 128); /* new address & 128 entries */
+//TRY   tstart = set_inch(uptr, mema, 128); /* new address & 128 entries */
```
Comment (pkoning2) [inline SEL32/sel32_con.c:354]: What is the purpose of this "TRY" line?  A note to you to try this in the future?  Leftover no longer needed code?

File: SEL32/sel32_con.c [hunk 3]
```diff
@@ -306,94 +363,21 @@ t_stat con_srvo(UNIT *uptr) {
...
         sim_debug(DEBUG_CMD, dptr,
-            "con_srvo INCH CMD %08x chsa %04x len %02x inch %06x\n", uptr->CMD, chsa, len, mema);
+            "con_srvo INCH CMD %08x chsa %04x len %02x inch %06x in2 %06x\n", uptr->CMD, chsa, len, mema, M[mema>>2]);
         /* WARNING, if SNS_DEVEND is not set, diags fail by looping in CON diag */
         chan_end(chsa, SNS_CHNEND|SNS_DEVEND);  /* return OK */
         break;
 
     case CON_NOP:       /* 0x03 */          /* NOP has do nothing */
-        uptr->CMD &= ~CON_MSK;              /* remove old CMD */
+//ZZ    uptr->CMD &= ~CON_MSK;              /* remove old CMD */
```
Comment (pkoning2) [inline SEL32/sel32_con.c:372]: ZZ?

File: SEL32/sel32_con.c [hunk 4]
```diff
@@ -669,7 +581,10 @@ t_stat con_srvi(UNIT *uptr) {
                 sim_debug(DEBUG_CMD, dptr,
                 "con_srvi readch unit %02x: CMD %08x read %02x u4 %02x incnt %02x\n",
                 unit, uptr->CMD, ch, uptr->u4, con_data[unit].incnt);
-            sim_activate(uptr, 30);         /* do this again */
+//JB        sim_clock_coschedule(uptr, 1000);
+//JB        sim_activate(uptr, 30);         /* do this again */
+//??        sim_activate(uptr, 300);        /* do this again */
+            sim_activate(uptr, 500);        /* do this again */
```
Comment (pkoning2) [inline SEL32/sel32_con.c:587]: Could you clean up all this commented out code everywhere?  It's hard to make sense out of any of them, and this spot with a half dozen nearly identical lines is particularly confusing.

File: SEL32/sel32_defs.h [hunk 5]
```diff
@@ -21,9 +21,99 @@
...
+#endif
+#else /* CPUONLY */
+#undef  USE_IPU_THREAD                  /* make sure IPU code undefined for CPUONLY */
+#undef DEFINE_IPU_MODELS                /* make sure IPU models undefine too */
+#endif /* CPU_ONLY */
+
+/* use correct variable type for thread IPU */
+#ifdef USE_IPU_THREAD
+#ifdef USE_IPU_CODE
+#define LOCAL static                    /* IPU in thread needs static variables */
```
Comment (pkoning2) [inline SEL32/sel32_defs.h:67]: I don't understand this.

"static" is about scope.  It has nothing to do with threading.  A variable defined at file scope has a single instance for the whole process, whether "static" or not.  There is such a thing as "thread local storage" (at least with some compilers and some platforms) but that isn't done with "static".

File: SEL32/sel32_chan.c (hunk 1 above)
Comment (AZBevier) [reply SEL32/sel32_chan.c:83]: These are comments to me about things I change during testing.  They usually go away during the next cleanup cycle.

File: SEL32/sel32_con.c (hunk 2 above)
Comment (AZBevier) [reply SEL32/sel32_con.c:354]: Try is again a debugging note to myself.  Will usually go away the next cleanup cycle.

File: SEL32/sel32_con.c (hunk 3 above)
Comment (AZBevier) [reply SEL32/sel32_con.c:372]: See above.

File: SEL32/sel32_con.c (hunk 4 above)
Comment (AZBevier) [reply SEL32/sel32_con.c:587]: Information to myself about various values I have tried during the past.  I try different timing for diags, UTX(UNIX), and MPX and they help me to remember what I tried.

File: SEL32/sel32_defs.h (hunk 5 above)
Comment (AZBevier) [reply SEL32/sel32_defs.h:67]: Bad choice for a comment.  All of the CPU/IPU registers are the same name.  Making the IPU names static allows the same code to reference different variables.  Not thread specific, but needed for IPU code.  I have lots of other changes I am making so all of these items can be addressed with the next push.  Lets merge these changes and move on.

Comment (pkoning2) [reply SEL32/sel32_defs.h:67]: Did you mean lots of changes on this topic, or more changes for other purposes?  In other words, is what we have here a complete feature/fix delivery?

File: SEL32/sel32_con.c (hunk 4 above)
Comment (pkoning2) [reply SEL32/sel32_con.c:587]: The source control system has history; it isn't necessary to keep it right in the source file active code.

**Comment by AZBevier** [comment]
Paul,

What is there now is a version well tested by me, but not others.  I was 
hoping for others to test the various versions too.  I have tested lots 
of UTX code, but not as much MPX-32 code, I was hoping for more.  I most 
likely will have more than one further release, two or three likely, to 
handle some enhancements and  lots of code cleanup and some timing tweaks.

Jim

**Comment by pkoning2** [comment]
> On Feb 7, 2023, at 8:27 PM, James C. Bevier ***@***.***> wrote:
> 
> 
> Paul,
> 
> What is there now is a version well tested by me, but not others.  I was 
> hoping for others to test the various versions too.  I have tested lots 
> of UTX code, but not as much MPX-32 code, I was hoping for more.  I most 
> likely will have more than one further release, two or three likely, to 
> handle some enhancements and  lots of code cleanup and some timing tweaks.

Thanks, that helps.  I'll go ahead and do the merge.

	paul

**Comment by pkoning2** [comment]
I just merged support for using CMake for building Open SIMH.

This was contributed by B. Scott Michel, who spent a lot of time and effort on this and responded to an amazing amount of feedback and comments.

Why CMake?  

It is a "meta-build" tool that allows a project such as Open SIMH to be built on many different platforms, using different tool sets, with a single set of build control files.  It also will allow things beyond simply building executable programs, for example creating distribution kit packages.

For now, CMake is an additional build mechanism.  That means Open SIMH now has three (Make, MSVC project files, and CMake).  Over time we would expect CMake to become the only one, with building on Unix type platforms (via make or other options like Ninja) and building on Windows with MSVC simply some of the things CMake can control.  At that point, the Cmake control files will be the ones that determine how Open SIMH is built.

Until then -- while we still have three build systems -- if you contribute changes that affect how one or more simulators need to be built, you should make sure all three sets of control files are updated properly.  For CMake this is automated; read the cmake related documentation that is part of Scott's contribution for details.  An example of when this is needed is if emulation of a new device is added to a simulator via a new source file.

	paul

**Pull request merged**